src/psyclone/domain/nemo/transformations/nemo_arrayaccess2loop_trans.py

```diff
@@ -162,31 +147,34 @@ def apply(self, node, options=None):
             loop_variable_symbol = DataSymbol(loop_variable_name, INTEGER_TYPE)
             symbol_table.add(loop_variable_symbol)
 
-        # Replace array access loop variable.
+        # Replace current access with loop variable.
         for array in assignment.walk(ArrayReference):
             if not array.ancestor(ArrayReference):
                 # This is not a nested access e.g. a(b(n)).
-                array.children[array_index] = Reference(loop_variable_symbol)
-
-        # As this transformation affects ancestor nodes of the
-        # supplied node we need to be careful here not to lose
-        # existing ancestor nodes as they might be
-        # referenced. Therefore use detach() and insert,() rather than
-        # copy() and replace_with().
-
-        # Find location.parent and location_index
-        loc_parent = location.parent
-        loc_index = location.position
+                array.indices[array_index] = Reference(loop_variable_symbol)
+
+        # Determine the loop body and where to add the loop.
+        nemo_kern = assignment.ancestor(NemoKern)
+        if nemo_kern:
+            # This assignment is inside a NemoKern
+            loop_body = nemo_kern
+        else:
+            # There is no parent NemoKern
+            loop_body = assignment
+        loc_parent = loop_body.parent
+        loc_index = loop_body.position
 
         # Create the new single-trip loop and add its children.
         step = Literal("1", INTEGER_TYPE)
         loop = NemoLoop.create(loop_variable_symbol, node_copy,
-                               node_copy.copy(), step, [location.detach()])
+                               node_copy.copy(), step, [loop_body.detach()])
+
         # Replace the original assignment with a loop containing the
         # modified assignment.
         loc_parent.children.insert(loc_index, loop)
 
-        if nloops == 0 and not assignment.walk(Range):
+        # Add a NemoKern if required.
+        if not nemo_kern and not assignment.walk(Range):
             # There were previously no loops therefore there was not
```

Comment:
This was not previously a NemoKern (as it contained no loops). However, we have now created a loop so, provided there are no range nodes, we must create an inlined kernel.

Reply:
I think you should write all the documentation :-)